examples/nlp/nl_image_search.py

```diff
@@ -0,0 +1,605 @@
+        )
+        # captions_similarity[i][j] is the dot_similarity(caption_i, caption_j).
+        captions_similarity = tf.matmul(
+            caption_embeddings, caption_embeddings, transpose_b=True
+        )
+        # targets[i][j] = avarage dot_similarity(caption_i, caption_j) and dot_similarity(image_i, image_j).
+        targets = keras.activations.softmax(
+            (captions_similarity + images_similarity) / (2 * self.temperature)
+        )
+        # Compute the loss for the captions using crossentropy
+        captions_loss = keras.losses.categorical_crossentropy(
+            y_true=targets, y_pred=logits, from_logits=True
+        )
+        # Compute the loss for the images using crossentropy
+        images_loss = keras.losses.categorical_crossentropy(
+            y_true=tf.transpose(targets), y_pred=tf.transpose(logits), from_logits=True
+        )
+        # Compute the average between the captions loss and the images loss
+        loss = (captions_loss + images_loss) / 2
+        # Return the mean of the loss over the batch.
+        return tf.math.reduce_mean(loss)
+
+
+"""
+## Implement the dual encoder
+"""
+
+
+def create_dual_encoder(text_encoder, vision_encoder):
+    # Receive the captions and the images as inputs.
+    captions = layers.Input(name="caption", shape=(), dtype=tf.string)
+    images = layers.Input(name="image", shape=(299, 299, 3), dtype=tf.float32)
+    # Get the embeddings for the captions.
+    caption_embeddings = text_encoder(captions)
+    # Get the embeddings for the images.
+    image_embeddings = vision_encoder(images)
+    # Stack the caption_embeddings and image_embeddings to produce
+    # a single output tensor for computing the loss.
+    outputs = tf.stack([caption_embeddings, image_embeddings])
+    # Create the dual encoder model.
+    dual_encoder = keras.Model(
+        inputs=[captions, images], outputs=outputs, name="dual_encoder"
+    )
+    return dual_encoder
+
+
+"""
+## Train the dual encoder model
+
+In this experiment, we freeze the base encoders for text and image, and make only
+the projection head trainable.
+"""
+
+projection_dims = 256
+num_projection_layers = 1
+dropout_rate = 0.1
+learning_rate = 0.003
+weight_decay = 0.0001
+num_epochs = 30
+batch_size = 256
+temperature = 0.07
+
+vision_encoder = create_vision_encoder(
+    num_projection_layers, projection_dims, dropout_rate
+)
+vision_encoder.summary()
+
+text_encoder = create_text_encoder(num_projection_layers, projection_dims, dropout_rate)
+text_encoder.summary()
+
+optimizer = tfa.optimizers.AdamW(learning_rate=learning_rate, weight_decay=weight_decay)
+loss = ContrastiveSimilarityLoss(temperature)
+dual_encoder = create_dual_encoder(text_encoder, vision_encoder)
+dual_encoder.compile(optimizer=optimizer, loss=loss)
+dual_encoder.summary()
+
+"""
+Note that training the model with 60,000 image-caption pairs takes around 15 minutes
+per epoch using a P100 GPU accelerator"""
```

Comment:
This is very dependent on the batch size, and potentially on the `steps_per_execution` parameter (how many batches to process in a single graph execution)